app.py

```diff
@@ -170,6 +170,13 @@ def before_request_proxy():
                     continue
                 request_headers [header [0]] = header [1]
 
+            # Sets session variables in a header x-session_vars
+            session_vars = {}
```

Comment:
Honestly, we probably need to get rid of session_vars that are only readable by a single instance anyway.

Sessions are currently stored on a single box, and this will never scale up to multiple Dynos (if requests get routed to different dynos then the session won't be there), same problem as with A/B testing.

So we need to either:

* Drop sessions and transport all the data via cookies; or
* Store session data in DynamoDB anyway.

Since our sessions are pretty small, we can probably do the first? Although if we have an  `is_admin` flag in the session, that might be trickier and we need something to verify the integrity of that cookie if we go that route...

Reply:
Good point. I didn't think of that either. I would rather store the sessions in DynamoDB so they can be retrieved everywhere. I'd much rather not have to trust cookies - in my experience, it's always best to give as little to the client in terms of state and let the server manage it.

In any case, the `is_admin` flag pertains to logged in users, and those only use sessions that are stored in DynamoDB. Those cookies are only identifiers and map to a username; the username then maps to its information and privileges.

I'm going to close the PR and implement logic for storing and retrieving session variables in DynamoDB.

Reply:
> in my experience, it's always best to give as little to the client in terms of state and let the server manage it.

I mean, of course, but consider this:

If we're going to store sessions in DynamoDB, every page request is going to require a GET to DynamoDB. This will increase latency and worker pool pressure (apparently GETs from outside EC2 aren't as fast as I thought they should be, looks like a DDB operation is routinely ~25-50ms), and it will increase DDB cost (we get charged for data as well as per-request). We will also accumulate garbage (every anonymous visitor will accrue a new record in DDB every time they visit the page... what will clean those up? Yes we can use TTL, now we need to set DDB TTL to match session lifetime...)

Now, those are the downsides. What are the upsides? What are the important pieces of information we store in the session that we are willing to pay these downsides for?

As far as I know, it's:

* A unique ID we use to tie programs together from anonymous users

Is there anything else I'm unaware of? Because if it's only the unique ID... it seems silly to send the users a unique ID in a cookie only to use that to retrieve a DIFFERENT unique ID from DynamoDB.

If the `is_admin` flag comes from DynamoDB anyway, and is not stored in a session, then there's no need for anything else.

So maybe the best solution is just to get rid of Flask session storage altogether, and just send the user a unique ID in a per-session cookie, say that is their "session id", and be done with it.

------------

> I'd much rather not have to trust cookies

Sure, but there are solutions for this. We can *sign* the cookies with a shared secret that's only known to the servers. That way, we can be sure that the server originally created the data we can detect if the user tried to mess with it (let's say they tried to maliciously flip their `is_admin` flag to `True`).

Keeping the server as stateless as possible is a good idea (for scaling), and having the client carry around a signed payload is a tried way of doing that without exposing yourself to risk.

Reply:
I'm aware that cookies can be signed and it's possible thus to trust client input.

Currently we don't need to store anything in the session, but this PR is meant to do that; an upcoming feature to be tested relies on session variables for non logged-in users. So we do potentially (and unfortunately) need to store session variables.

Regarding the performance, we currently already do a request to DynamoDB per every request for logged in users, to resolve their username from their session. Storing the sessions for anonymous requests would multiply the load by `1 / proportion of requests done by logged in users`. This could be quite significant, you're right about that. However, I believe we chose DynamoDB because it's fast and cheap, so we shouldn't be deterred from using it.

Maybe it's best, as you suggested, to store the session id (and other possible variables) in a signed cookie. We could furthermore delete that session cookie when the user logs out, to make sure that it doesn't remain behind in case sensitive data is stored in there. I'll start working in that direction.

Reply:
> we chose DynamoDB because it's fast and cheap

Not as fast as I'm used to, at least not from Heroku 😅